src/commands/index.rs

```diff
@@ -0,0 +1,163 @@
+use serde_json::Value;
+
+use redismodule::{Context, RedisError, RedisResult, RedisValue};
+use redismodule::{NextArg, REDIS_OK};
+
+use redisearch_api::{Document, FieldType};
+
+use crate::error::Error;
+use crate::redisjson::{Format, RedisJSON};
+use crate::schema::Schema;
+use crate::REDIS_JSON_TYPE;
+
+pub mod schema_map {
+    use crate::schema::Schema;
+    use std::collections::HashMap;
+
+    type SchemaMap = HashMap<String, Schema>;
+
+    /// We keep a static map, since it needs to be accessed from multiple Redis module commands
+    /// and there is no other obvious way to implement this (such as a "user data" pointer maintained
+    /// for us by Redis).
+    ///
+    /// The init function should be called only once.
+    /// Unwrapping the Option is thus safe afterwards.
+    ///
+    /// Since we have only one thread at the moment, getting a (mutable) reference to the map
+    /// is also safe. If we add threads, we can simply wrap the map: `Option<Mutex<SchemaMap>`.
+    ///
+    static mut SCHEMA_MAP: Option<SchemaMap> = None;
+
+    pub fn init() {
+        let map = HashMap::new();
+        unsafe {
+            SCHEMA_MAP = Some(map);
+        }
+    }
+
+    pub fn as_ref() -> &'static SchemaMap {
+        unsafe { SCHEMA_MAP.as_ref() }.unwrap()
+    }
+
+    pub fn as_mut() -> &'static mut SchemaMap {
+        unsafe { SCHEMA_MAP.as_mut() }.unwrap()
+    }
+}
+
+///////////////////////////
+
+fn add_field<I>(index_name: &str, field_name: &str, mut args: I) -> RedisResult
+where
+    I: Iterator<Item = String>,
+{
+    let path = args.next_string()?;
+
+    let map = schema_map::as_mut();
+
+    if !map.contains_key(index_name) {
+        let schema = Schema::new(&index_name);
+        map.insert(index_name.to_owned(), schema);
+    }
+
+    let schema = map.get_mut(index_name).unwrap();
```

Comment:
Isn't that a waste to call get_mut if you just created the schema in the "if" above?

Reply:
It seems to be, but I don't have much choice: the `schema` value is consumed by the `map.insert` above (the `HashMap` owns it), and therefore it can no longer be used.

Reply:
so at least let's do something like that to avoid call get_mut and contains_key on every path
```
    let schema = if let Some(stored_schema) = map.get_mut(index_name){
        stored_schema
    } else {
        let new_schema = Schema::new(&index_name);
        map.insert(index_name.to_owned(), new_schema);
        map.get_mut(index_name).unwrap()
    };
```

Reply:
or something like that 
```
    let schema = map
        .entry(index_name.to_owned())
        .or_insert_with(|| Schema::new(&index_name));
```
And we can even avoid to_owned() if you pass owned String and not &str to add_field

Reply:
Good suggestion, done.